State Triggers (#351)
MicrosoftDocs/WindowsCommunityToolkitDocs: docs/helpers/Triggers.md

---
title: State Triggers
author: dotMorten
description: A collection of custom visual State Triggers
keywords: windows 10, uwp, windows community toolkit, uwp community toolkit, uwp toolkit, CompareStateTrigger, FullScreenModeStateTrigger, IsEqualStateTrigger, IsNotEqualStateTrigger, IsNullOrEmptyStateTriggers, NetworkConnectionStateTrigger, RegexStateTrigger, UserHandPreferenceStateTrigger, UserInteractionModeStateTrigger

---

# State Triggers

<!-- Describe your control -->

> A collection of custom visual [State Triggers](https://docs.microsoft.com/dotnet/api/Microsoft.Toolkit.Uwp.UI.Triggers)

Reviewer: In this case, let's actually link to the system api docs as that's the base class (and each has a link below to our api doc)

```suggestion
A collection of custom visual [State Triggers](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.statetrigger).
```

> <!-- Use below format to display note
> > [!NOTE]
> > Some note
> 
> > [!IMPORTANT]
> > Some important note
> 
> > [!WARNING]
> > Some warning note
> -->
> 

Reviewer: We can remove the template bits we're not using

```suggestion
```

| [IsNullOrEmptyStateTriggers](https://docs.microsoft.com/dotnet/api/Microsoft.Toolkit.Uwp.UI.Triggers.IsNullOrEmptyStateTriggers) | Enables a state if an Object is null or a String/IEnumerable is empty |
| [NetworkConnectionStateTrigger](https://docs.microsoft.com/dotnet/api/Microsoft.Toolkit.Uwp.UI.Triggers.NetworkConnectionStateTrigger) | Trigger for switching when the network availability changes |
| [RegexStateTrigger](https://docs.microsoft.com/dotnet/api/Microsoft.Toolkit.Uwp.UI.Triggers.RegexStateTrigger) | Enables a state if the regex expression is true for a given string value |
| [UserHandPreferenceStateTrigger](https://docs.microsoft.com/dotnet/api/Microsoft.Toolkit.Uwp.UI.Triggers.UserHandPreferenceStateTrigger) | Trigger for switching UI based on whether the user favors their left or right hand |
| [UserInteractionModeStateTrigger](https://docs.microsoft.com/dotnet/api/Microsoft.Toolkit.Uwp.UI.Triggers.UserInteractionModeStateTrigger) | Trigger for switching when the User interaction mode changes (tablet mode) |

## CompareStateTrigger Example

> ```

Reviewer: If we add a code-fence tag for XML this should get syntax highlighting in the docs.

```suggestion
```xml
```

<VisualState.StateTriggers>

>     <triggers:CompareStateTrigger Value="{Binding Value,ElementName=Slider, Mode=OneWay}" To="3" Comparison="LessThan" />

Reviewer: Just reads a bit clearer like an equation this way:

```suggestion
    <triggers:CompareStateTrigger Value="{Binding Value,ElementName=Slider, Mode=OneWay}" Comparison="LessThanOrEqual" To="3"/>
```

</VisualState.StateTriggers>
```
## FullScreenModeStateTrigger Example 

> ```

Reviewer: ```suggestion
```xml
```

<VisualState.StateTriggers>
    <triggers:FullScreenModeStateTrigger IsFullScreen="true" />
</VisualState.StateTriggers>
```                    
## IsEqualStateTrigger Example

> ```

Reviewer: ```suggestion
```xml
```

<VisualState.StateTriggers>
    <triggers:IsEqualStateTrigger Value="{Binding IsChecked, ElementName=checkbox, Mode=OneWay}" To="{x:Null}" />
</VisualState.StateTriggers>
```
## IsNotEqualStateTrigger Example

> ```

Reviewer: ```suggestion
```xml
```

<VisualState.StateTriggers>
    <triggers:IsNotEqualStateTrigger Value="{Binding IsChecked, ElementName=checkbox, Mode=OneWay}" To="{x:Null}" />
</VisualState.StateTriggers>
```
## IsNullOrEmptyStateTriggers Example

> ```

Reviewer: ```suggestion
```xml
```

<VisualState.StateTriggers>
    <triggers:IsNullOrEmptyStateTrigger Value="{Binding Text, ElementName=OurTextBox, Mode=OneWay}"/>
</VisualState.StateTriggers>
```              
## NetworkConnectionStateTrigger Example

> ```

Reviewer: ```suggestion
```xml
```

<VisualState.StateTriggers>
    <triggers:NetworkConnectionStateTrigger ConnectionState="Connected" />
</VisualState.StateTriggers>
```
## RegexStateTrigger Example

> ```

Reviewer: ```suggestion
```xml
```

<VisualState.StateTriggers>

>     <triggers:RegexStateTrigger Value="{Binding Text, ElementName=emailTextBox, Mode=OneWay}"
> 		                    Expression="^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$"
> 				    Options="IgnoreCase" />

Reviewer: Think there's a mix of tabs and spaces here that was causing issues with alignment in the page:

```suggestion
    <triggers:RegexStateTrigger Value="{Binding Text, ElementName=emailTextBox, Mode=OneWay}"
                                                  Expression="^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$"
                                                  Options="IgnoreCase" />
```

</VisualState.StateTriggers>
```
## UserHandPreferenceStateTrigger Example

> ```

Reviewer: ```suggestion
```xml
```

<VisualState.StateTriggers>
    <triggers:UserHandPreferenceStateTrigger HandPreference="LeftHanded" />
</VisualState.StateTriggers>
```
## UserInteractionModeStateTrigger Example

> ```

Reviewer: ```suggestion
```xml
```

<VisualState.StateTriggers>
    <triggers:UserInteractionModeStateTrigger InteractionMode="Mouse" />
</VisualState.StateTriggers>
```

## Sample Project

<!-- Link to the sample page in the Windows Community Toolkit Sample App -->
[Triggers sample page Source](https://github.com/windows-toolkit/WindowsCommunityToolkit/tree/master/Microsoft.Toolkit.Uwp.SampleApp/SamplePages/Triggers). You can see this in action in the [Windows Community Toolkit Sample App](http://aka.ms/uwptoolkitapp).

## Requirements

| Device family | Universal, 10.0.16299.0 or higher   |
| -- | -- |
| Namespace | Microsoft.Toolkit.Uwp.UI.Triggers |

> | NuGet package | [Microsoft.Toolkit.Uwp](https://www.nuget.org/packages/Microsoft.Toolkit.Uwp.UI/) |

Reviewer: ```suggestion
| NuGet package | [Microsoft.Toolkit.Uwp.UI](https://www.nuget.org/packages/Microsoft.Toolkit.Uwp.UI/) |
```
